// modern-face-api/build/commonjs/mtcnn/extractParamsFromWeigthMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractParamsFromWeigthMap = extractParamsFromWeigthMap;
const common_1 = require("../common");
function extractorsFactory(weightMap, paramMappings) {
    const extractWeightEntry = (0, common_1.extractWeightEntryFactory)(weightMap, paramMappings);
    function extractConvParams(prefix) {
        const filters = extractWeightEntry(`${prefix}/weights`, 4, `${prefix}/filters`);
        const bias = extractWeightEntry(`${prefix}/bias`, 1);
        return { filters, bias };
    }
    function extractFCParams(prefix) {
        const weights = extractWeightEntry(`${prefix}/weights`, 2);
        const bias = extractWeightEntry(`${prefix}/bias`, 1);
        return { weights, bias };
    }
    function extractPReluParams(paramPath) {
        return extractWeightEntry(paramPath, 1);
    }
    function extractSharedParams(prefix) {
        const conv1 = extractConvParams(`${prefix}/conv1`);
        const prelu1_alpha = extractPReluParams(`${prefix}/prelu1_alpha`);
        const conv2 = extractConvParams(`${prefix}/conv2`);
        const prelu2_alpha = extractPReluParams(`${prefix}/prelu2_alpha`);
        const conv3 = extractConvParams(`${prefix}/conv3`);
        const prelu3_alpha = extractPReluParams(`${prefix}/prelu3_alpha`);
        return { conv1, prelu1_alpha, conv2, prelu2_alpha, conv3, prelu3_alpha };
    }
    function extractPNetParams() {
        const sharedParams = extractSharedParams('pnet');
        const conv4_1 = extractConvParams('pnet/conv4_1');
        const conv4_2 = extractConvParams('pnet/conv4_2');
        return Object.assign(Object.assign({}, sharedParams), { conv4_1, conv4_2 });
    }
    function extractRNetParams() {
        const sharedParams = extractSharedParams('rnet');
        const fc1 = extractFCParams('rnet/fc1');
        const prelu4_alpha = extractPReluParams('rnet/prelu4_alpha');
        const fc2_1 = extractFCParams('rnet/fc2_1');
        const fc2_2 = extractFCParams('rnet/fc2_2');
        return Object.assign(Object.assign({}, sharedParams), { fc1, prelu4_alpha, fc2_1, fc2_2 });
    }
    function extractONetParams() {
        const sharedParams = extractSharedParams('onet');
        const conv4 = extractConvParams('onet/conv4');
        const prelu4_alpha = extractPReluParams('onet/prelu4_alpha');
        const fc1 = extractFCParams('onet/fc1');
        const prelu5_alpha = extractPReluParams('onet/prelu5_alpha');
        const fc2_1 = extractFCParams('onet/fc2_1');
        const fc2_2 = extractFCParams('onet/fc2_2');
        const fc2_3 = extractFCParams('onet/fc2_3');
        return Object.assign(Object.assign({}, sharedParams), { conv4, prelu4_alpha, fc1, prelu5_alpha, fc2_1, fc2_2, fc2_3 });
    }
    return {
        extractPNetParams,
        extractRNetParams,
        extractONetParams,
    };
}
function extractParamsFromWeigthMap(weightMap) {
    const paramMappings = [];
    const { extractPNetParams, extractRNetParams, extractONetParams } = extractorsFactory(weightMap, paramMappings);
    const pnet = extractPNetParams();
    const rnet = extractRNetParams();
    const onet = extractONetParams();
    (0, common_1.disposeUnusedWeightTensors)(weightMap, paramMappings);
    return { params: { pnet, rnet, onet }, paramMappings };
}
