// modern-face-api/build/commonjs/mtcnn/MtcnnOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MtcnnOptions = void 0;
class MtcnnOptions {
    constructor({ minFaceSize, scaleFactor, maxNumScales, scoreThresholds, scaleSteps, } = {}) {
        this._name = 'MtcnnOptions';
        this._minFaceSize = minFaceSize || 20;
        this._scaleFactor = scaleFactor || 0.709;
        this._maxNumScales = maxNumScales || 10;
        this._scoreThresholds = scoreThresholds || [0.6, 0.7, 0.7];
        this._scaleSteps = scaleSteps;
        if (typeof this._minFaceSize !== 'number' || this._minFaceSize < 0) {
            throw new Error(`${this._name} - expected minFaceSize to be a number > 0`);
        }
        if (typeof this._scaleFactor !== 'number' || this._scaleFactor <= 0 || this._scaleFactor >= 1) {
            throw new Error(`${this._name} - expected scaleFactor to be a number between 0 and 1`);
        }
        if (typeof this._maxNumScales !== 'number' || this._maxNumScales < 0) {
            throw new Error(`${this._name} - expected maxNumScales to be a number > 0`);
        }
        if (!Array.isArray(this._scoreThresholds) ||
            this._scoreThresholds.length !== 3 ||
            this._scoreThresholds.some(th => typeof th !== 'number')) {
            throw new Error(`${this._name} - expected scoreThresholds to be an array of numbers of length 3`);
        }
        if (this._scaleSteps &&
            (!Array.isArray(this._scaleSteps) || this._scaleSteps.some(th => typeof th !== 'number'))) {
            throw new Error(`${this._name} - expected scaleSteps to be an array of numbers`);
        }
    }
    get minFaceSize() {
        return this._minFaceSize;
    }
    get scaleFactor() {
        return this._scaleFactor;
    }
    get maxNumScales() {
        return this._maxNumScales;
    }
    get scoreThresholds() {
        return this._scoreThresholds;
    }
    get scaleSteps() {
        return this._scaleSteps;
    }
}
exports.MtcnnOptions = MtcnnOptions;
