// modern-face-api/build/commonjs/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTensor = isTensor;
exports.isTensor1D = isTensor1D;
exports.isTensor2D = isTensor2D;
exports.isTensor3D = isTensor3D;
exports.isTensor4D = isTensor4D;
exports.isFloat = isFloat;
exports.isEven = isEven;
exports.round = round;
exports.isDimensions = isDimensions;
exports.computeReshapedDimensions = computeReshapedDimensions;
exports.getCenterPoint = getCenterPoint;
exports.range = range;
exports.isValidNumber = isValidNumber;
exports.isValidProbablitiy = isValidProbablitiy;
const tf = require("@tensorflow/tfjs-core");
const classes_1 = require("../classes");
const Dimensions_1 = require("../classes/Dimensions");
function isTensor(tensor, dim) {
    return tensor instanceof tf.Tensor && tensor.shape.length === dim;
}
function isTensor1D(tensor) {
    return isTensor(tensor, 1);
}
function isTensor2D(tensor) {
    return isTensor(tensor, 2);
}
function isTensor3D(tensor) {
    return isTensor(tensor, 3);
}
function isTensor4D(tensor) {
    return isTensor(tensor, 4);
}
function isFloat(num) {
    return num % 1 !== 0;
}
function isEven(num) {
    return num % 2 === 0;
}
function round(num, prec = 2) {
    const f = 10 ** prec;
    return Math.floor(num * f) / f;
}
function isDimensions(obj) {
    return !!(obj &&
        typeof obj === 'object' &&
        obj.width &&
        obj.height);
}
function computeReshapedDimensions({ width, height }, inputSize) {
    const scale = inputSize / Math.max(height, width);
    return new Dimensions_1.Dimensions(Math.round(width * scale), Math.round(height * scale));
}
function getCenterPoint(pts) {
    return pts
        .reduce((sum, pt) => sum.add(pt), new classes_1.Point(0, 0))
        .div(new classes_1.Point(pts.length, pts.length));
}
function range(num, start, step) {
    return Array(num)
        .fill(0)
        .map((_, i) => start + i * step);
}
function isValidNumber(num) {
    return (!!num && num !== Infinity && num !== -Infinity && !Number.isNaN(num)) || num === 0;
}
function isValidProbablitiy(num) {
    return isValidNumber(num) && typeof num === 'number' && 0 <= num && num <= 1.0;
}
